namer/namer.cc

```diff
@@ -1085,6 +1150,34 @@ class SymbolDefiner {
         }
     }
 
+    core::FieldRef insertField(core::MutableContext ctx, const core::FoundField &field) {
+        auto scope = ctx.owner.enclosingClass(ctx);
+
+        auto existing = scope.data(ctx)->findMember(ctx, field.name);
```

Comment:
In resolver, the code to enter class variables is preceded by some logic to make sure if the user writes

```ruby
class A
  def example
    @@x = T.let(0, T.nilable(Integer))
  end
end
```

that an error is reported and then the owner is changed to the enclosing class.

Should we move that logic here into namer?

Since that logic is present in resolver but absent from namer, I'm worried that we might be entering the `@@x` variables declared inside methods (both instance and singleton) on in two separate places.

Reply:
We have tests for what happens when we incorrectly declare variables in the wrong context (`test/testdata/resolver/let_errors.rb`, including a symbol table exp test), and those tests did not change as a result of this patch series, so we can at least say that we're not putting things in the wrong place.  (It is possible that future changes would de-sync namer and resolver and/or remove the test coverage that we had, so putting the logic in a single place might be a good idea anyway.)

That error-reporting logic is a little awkward to port over with the last few changes, because that logic is dependent upon the ownership of `ctx` being correctly attributed to a method, which would argue in favor of fields being "owned" by methods.  I will see if the past few changes can be reverted and the errors moved.